stage1/init/common/pod.go

```diff
@@ -276,6 +262,116 @@ func findHostPort(pm schema.PodManifest, name types.ACName) uint {
+	sysusersConf = append(sysusersConf, fmt.Sprintf("u %s %d \"%s\"\n", username, uid_, username))
+
+	if err := ioutil.WriteFile(path.Join(common.Stage1RootfsPath(p.Root), "usr/lib/sysusers.d", ServiceUnitName(appName)+".conf"),
+		[]byte(strings.Join(sysusersConf, "\n")), 0640); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// lookupPathInsideApp returns the path (relative to the app rootfs) of the
+// given binary. It will look up on "paths" (also relative to the app rootfs)
+// and evaluate possible symlinks to check if the resulting path is actually
+// executable.
+func lookupPathInsideApp(bin string, paths string, appRootfs string) (string, error) {
+	pathsArr := filepath.SplitList(paths)
+	var appPathsArr []string
+	for _, p := range pathsArr {
+		appPathsArr = append(appPathsArr, filepath.Join(appRootfs, p))
+	}
+	for _, path := range appPathsArr {
+		binPath := filepath.Join(path, bin)
+		binAbsPath, err := filepath.Abs(binPath)
+		if err != nil {
+			return "", fmt.Errorf("unable to find absolute path for %s", binPath)
+		}
+		relPath := strings.TrimPrefix(binAbsPath, appRootfs)
+		d, err := os.Lstat(binAbsPath)
+		if err != nil {
+			continue
+		}
+		binRealPath, err := evaluateSymlinksInsideApp(appRootfs, relPath)
+		if err != nil {
+			return "", errwrap.Wrap(fmt.Errorf("could not evaluate path %v", binAbsPath), err)
+		}
+		binRealPath = filepath.Join(appRootfs, binRealPath)
+		d, err = os.Stat(binRealPath)
+		if err != nil {
+			continue
+		}
+		// Check the executable bit, inspired by os.exec.LookPath()
+		if m := d.Mode(); !m.IsDir() && m&0111 != 0 {
+			// The real path is executable, return the path relative to the app
+			return relPath, nil
+		}
+	}
+	return "", fmt.Errorf("unable to find %q in %q", bin, paths)
+}
+
+// appSearchPaths returns a list of paths where we should search for
+// non-absolute exec binaries
+func appSearchPaths(p *stage1commontypes.Pod, appRootfs string, app types.App) []string {
+	appEnv := app.Environment
+
+	if imgPath, ok := appEnv.Get("PATH"); ok {
+		return strings.Split(imgPath, ":")
+	}
+
+	// emulate exec(3) behavior, first check pod's directory and then the list
+	// of directories returned by confstr(_CS_PATH). That's typically
+	// "/bin:/usr/bin" so let's use that.
+	return []string{appRootfs, "/bin", "/usr/bin"}
+}
+
+// findBinPath takes a binary path and returns a the absolute path of the
+// binary relative to the app rootfs. This can be passed to ExecStart on the
+// app's systemd service file directly.
+func findBinPath(p *stage1commontypes.Pod, appName types.ACName, app types.App, bin string) (string, error) {
+	absRoot, err := filepath.Abs(p.Root)
+	if err != nil {
+		return "", errwrap.Wrap(errors.New("could not get pod's root absolute path"), err)
+	}
+
+	appRootfs := common.AppRootfsPath(absRoot, appName)
+	appPathDirs := appSearchPaths(p, appRootfs, app)
+	appPath := strings.Join(appPathDirs, ":")
+
+	binPath, err := lookupPathInsideApp(bin, appPath, appRootfs)
```

Comment:
`bin` can have 3 different kinds of values:
- an absolute path (starting with "/", such as "/bin/sh"). In this case, lookupPathInsideApp should not be called.
- a filename that does not contain a slash (/) character (such as "sh"). In this case, appSearchPaths should be called.
- a relative path (such as "data/run.sh"). In this case, it should not look in $PATH but only in the working directory as defined by the ACI manifest.